lib/src/annotate.rs

```diff
@@ -0,0 +1,382 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Methods that allow annotation (attribution and blame) for a file in a
+//! repository.
+//!
+//! TODO: Add support for different blame layers with a trait in the future.
+//! Like commit metadata and more.
+
+use std::collections::HashMap;
+
+use pollster::FutureExt;
+use thiserror::Error;
+
+use crate::backend::{BackendError, CommitId};
+use crate::commit::Commit;
+use crate::conflicts::{materialize_tree_value, MaterializedTreeValue};
+use crate::diff::{Diff, DiffHunk};
+use crate::fileset::FilesetExpression;
+use crate::graph::{GraphEdge, GraphEdgeType};
+use crate::merged_tree::MergedTree;
+use crate::repo::Repo;
+use crate::repo_path::RepoPath;
+use crate::revset::{RevsetEvaluationError, RevsetExpression, RevsetFilterPredicate};
+use crate::store::Store;
+
+/// Various errors that can arise from annotation
+#[derive(Debug, Error)]
+pub enum AnnotateError {
+    /// the requested file path was not found
+    #[error("Unable to locate file: {0}")]
+    FileNotFound(String),
+    /// the file type is incorrect. Usually a directory was given but a regular
+    /// file is required
+    #[error("File {0} must be a regular file, not a directory")]
+    UnsupportedFileType(String),
+    /// the file is in a conflicted state and can therefore not be annotated
+    /// properly
+    #[error("File {0} is conflicted at commit: {1}")]
+    Conflicted(String, String),
```

Comment:
nit: `UnsupportedFileType` and `Conflicted` are no longer used. Maybe `FileNotFound` can also replaced with an empty result.

Reply:
Removing the unused errors. I'm hesitant about removing FileNotFound because this is a library function. 
From the UI, it already checks that the file exists, but, if other functions call this with a non-existent file, I think it makes sense to throw an error rather than empty results right?

Reply:
We already special case the starting path in the cli component (if it doesn't exist we error out before going to jj_lib). 
With that in mind, I've taken your suggestion and removed FileNotFound (returning empty results if we can't find the starting file) and therefore AnnotateError in general (since there was nothing left). It now returns a BackendError like other functions in jj_lib.

I also integrated your latest changes to materialize_merge_results which impacted my code. Take a look and let me know if I integrated your changes correctly.